_posts/2018-04-11-sign-to-contract.md

```diff
@@ -0,0 +1,400 @@
+obtaining the signed (and broadcastable) transaction `TX`.
+Then sends `TX` to the network and, later on, when confirmed (~6 blocks), it becomes part of the blockchain forever.
+
+We can see that inside `ECDSAsig` an elliptic curve point (`R`) is used.
+The idea of *sign-to-contract* is simply to tweak it with `h(R||c)`, 
+making it __also__ a commitment to another value, the contract `c`.
+
+```
+def ECDSAsign2contract(x,m,c):
+    k = deterministic_nonce(x,m)
+    R = k*G
+    e = k + h(R||c)
+    Q = R + h(R||c)*G  # which is Q = e*G = C(c,R)
+    q = Q.x mod n
+    z = e^(-1) * (m + q*x) mod n
+    return (q,z), R
+```
+Carl applies `ECDSAsign2contract`, which produces `(q,z)` and `R`. 
+The former is the signature corresponding to the pubkey `x*G` for the message `m`,
+while the latter and `c` are the ingredients to prove the commitment to `Q.x`. 
+Now `uTX` is filled with `SIG(q,z)` which contains `q`, which in turn is (almost always) `Q.x`.
+As before, the resulting `TX` then is committed to the blockchain.
+
+Carl then looks at `TX`, in which he spots `Q.x`, with `TX = TXp||Q.x||TXa`. 
+By retrieving the info to link `TX` to a block header we can complete the timestamp for the contract `c`:
+```
+File sha256: c
+Timestamp:
+prepend R == R||c
+secp256k1commitment == Q.x
+prepend TXp == TXp||Q.x
+append TXa == TXp||Q.x||TXa
+# transaction id ...
+sha256
+...
+sha256
+verify BitcoinBlockHeaderAttestation(...)
+# Bitcoin merkle root ...
+``` 
+**Resuming**, while doing a purely financial transaction to Diana, **Carl timestamped** a contract, 
+without adding any bytes, so **with zero marginal cost**.
+Diana could be unaware that Carl timestamped, 
+from her prospective the coins she just received are indistinguishable from coins signed in the standard way. 
+Such coins are locked in the pubkey (or address) that Diana told Carl, 
+hence Diana's new coins are as secured as normal coins. 
+
+But what about Carl?
+Aside from timestamping, can this procedure compromise his security?
+
+The delicate phase is between the broadcast and the confirmation, 
+if, in that period, Carl secret key `x` gets exposed, an eavesdropper, Eve, will redirect Carl's coin to her. 
+However, once the transaction to Diana is confirmed, Eve cannot do much (unless Carl does address reuse), 
+since the coins locked by `x` are already spent.
+The private key can also be revealed in a indirect way, 
+in fact in a signature revealing the nonce `k` 
+(or `e` to be consistent with the above formula) 
+is equivalent to expose `x`.
+Still *sign-to-contract* does not increase the risks Carl faces with a standard signature.
+Indeed when showing his OTS proof he just claims that what can be found  in the signature (namely `Q.x`) was obtained from the elliptic curve point (`Q=e*G`) which is the sum of two other specific points (`R` and `h(R||c)*G`),
+but actually this does not give any additional clue to solve the discrete logarithm that secures the signature.
```

Comment:
I think this part is confusing to the reader. You're talking about secret key compromise in a way that doesn't make it clear that it's not an issue.